server/src/main/java/org/elasticsearch/action/admin/indices/retention/TransportSyncRetentionLeasesAction.java

```diff
@@ -78,7 +79,12 @@ protected SyncRetentionLeasesRequest readRequestFrom(StreamInput in) throws IOEx
     @Override
     protected void shardOperation(SyncRetentionLeasesRequest request, ShardRouting shardRouting, ActionListener<ReplicationResponse> listener) throws IOException {
         IndexShard indexShard = indicesService.indexServiceSafe(shardRouting.shardId().getIndex()).getShard(shardRouting.shardId().id());
-        indexShard.syncRetentionLeases(true, listener);
+        indexShard.runUnderPrimaryPermit(
+            () -> indexShard.syncRetentionLeases(true, listener),
+            e -> logger.warn(new ParameterizedMessage("failed to sync retention leases on OPTIMIZE TABLE"), e),
```

Comment:
I think we need to wrap this exception in a BroadcastShardOperationFailedException and pass it to `listener.onFailure()`, otherwise the request won't ever get completed if a primary permit can't be acquired.

Reply:
listener here is a wrapped one (https://github.com/crate/crate/commit/8c0b43069e9460f84d0efad09b368128962381ba), and it takes care of wrapping the exception in a `BroadcastShardOperationFailedException` onFailure - should be fine or I;m missing something?

Reply:
Ah yes, sorry hadn't updated my local copy!  We still need to pass `e` to the listener I think?

Reply:
I deleted comment (I thought it's invalid), getting it back

>listener here is a wrapped one (https://github.com/crate/crate/commit/8c0b43069e9460f84d0efad09b368128962381ba), and it takes care of wrapping the exception in a BroadcastShardOperationFailedException - should be fine or I'm missing something?

>We still need to pass e to the listener I think?

Right, `runUnderPrimaryPermit` wraps (again) our listener and overrides exception conversion logic we need by by the Consumer. Looks like need to duplicate exception wrapping(

Reply:
Thanks for catching, passed exception to the listener and added a comment